pkg/graveler/ref/branch_iterator.go

```diff
@@ -2,143 +2,166 @@ package ref
-		buf:             make([]*graveler.BranchRecord, 0, prefetchSize),
-		orderByCommitID: false,
+func (bi *BranchSimpleIterator) Next() bool {
+	if bi.Err() != nil {
+		return false
 	}
-	for _, opt := range opts {
-		opt(res)
+	if !bi.itr.Next() {
+		bi.value = nil
+		return false
 	}
-	return res
-}
-
-func (ri *BranchIterator) Next() bool {
-	if ri.err != nil {
+	entry := bi.itr.Entry()
+	if entry == nil {
+		bi.err = graveler.ErrInvalid
 		return false
 	}
-	ri.maybeFetch()
-
-	// stage a value and increment offset
-	if len(ri.buf) == 0 {
+	value, ok := entry.Value.(*graveler.BranchData)
+	if !ok {
+		bi.err = graveler.ErrReadingFromStore
 		return false
 	}
-	ri.value = ri.buf[0]
-	ri.buf = ri.buf[1:]
-	ri.offset = string(ri.value.BranchID)
+
+	bi.value = &graveler.BranchRecord{
+		BranchID: graveler.BranchID(value.Id),
+		Branch:   branchFromProto(value),
+	}
 	return true
 }
 
-func (ri *BranchIterator) maybeFetch() {
-	if ri.state == iteratorStateDone {
-		return
+func (bi *BranchSimpleIterator) SeekGE(id string) {
+	if bi.Err() == nil {
+		bi.itr.Close() // Close previous before creating new iterator
+		bi.itr, bi.err = kv.NewPrimaryIterator(bi.ctx, bi.store.Store, (&graveler.BranchData{}).ProtoReflect().Type(),
+			bi.repoPartition, []byte(graveler.BranchPath("")), kv.IteratorOptionsFrom([]byte(graveler.BranchPath(graveler.BranchID(id)))))
 	}
-	if len(ri.buf) > 0 {
-		return
+}
+
+func (bi *BranchSimpleIterator) Value() *graveler.BranchRecord {
+	if bi.Err() != nil {
+		return nil
 	}
+	return bi.value
+}
 
-	var offsetCondition string
-	if ri.state == iteratorStateInit {
-		offsetCondition = iteratorOffsetCondition(true)
-		ri.state = iteratorStateQuerying
-	} else {
-		offsetCondition = iteratorOffsetCondition(false)
+func (bi *BranchSimpleIterator) Err() error {
+	if bi.err != nil {
+		return bi.err
 	}
+	return bi.itr.Err()
+}
+
+func (bi *BranchSimpleIterator) Close() {
+	bi.err = ErrIteratorClosed
+	bi.itr.Close()
+}
+
+type BranchByCommitIterator struct {
+	ctx    context.Context
+	values []*graveler.BranchRecord
+	value  *graveler.BranchRecord
+	idx    int
+	err    error
+}
+
+func (b *BranchByCommitIterator) SortByCommitID(i, j int) bool {
+	return b.values[i].CommitID.String() <= b.values[j].CommitID.String()
+}
 
-	var buf []*branchRecord
-	err := ri.db.Select(ri.ctx, &buf, `
-			SELECT id, staging_token, commit_id
-			FROM graveler_branches
-			WHERE repository_id = $1
-			AND id `+offsetCondition+` $2
-			ORDER BY `+ri.getOrderBy()+`  ASC
-			LIMIT $3`, ri.repositoryID, ri.offset, ri.fetchSize)
+func NewBranchByCommitIterator(ctx context.Context, store *kv.StoreMessage, repo *graveler.RepositoryRecord) (*BranchByCommitIterator, error) {
+	bi := &BranchByCommitIterator{
+		ctx:    ctx,
+		values: make([]*graveler.BranchRecord, 0),
+	}
+	itr, err := kv.NewPrimaryIterator(ctx, store.Store, (&graveler.BranchData{}).ProtoReflect().Type(), graveler.RepoPartition(repo), []byte(""), kv.IteratorOptionsFrom([]byte("")))
```

Comment:
Are you trying to get all the branches? I think you're missing the branch prefix..
I believe it would be much simpler to use the `BranchSimpleIterator` here, it's basically what it does..